nwbwidgets/timeseries.py

```diff
@@ -966,4 +942,485 @@ def update(
+    return data.flatten()
+
+
+def trialize_time_series(
+    time_series: TimeSeries,
+    trials_table: Union[DynamicTable, pd.DataFrame],
+    data_column: Optional[int] = 0,
+    start_time_shift: float = 0,
+    duration: float = 1,
+    alignment_column: str = "start_time",
+) -> pd.DataFrame:
+
+    if isinstance(trials_table, DynamicTable):
+        trials_table_df = trials_table.to_dataframe()
+    else:
+        trials_table_df = trials_table
+
+    trials_table_df = trials_table_df.reset_index().rename(columns={"id": "trial"})
+
+    # Map timestamps to the trial interval (start_time, stop_time)
+    timestamps = get_timeseries_tt(node=time_series)
+    # Get
+    values_to_align_array = trials_table_df[alignment_column].to_numpy()
+    left_trial_bound = values_to_align_array + start_time_shift
+    left_trial_bound = left_trial_bound[
+        left_trial_bound <= timestamps.max()
+    ]  # Filter values outside of the time series to avoid extra computation
+    right_trial_bound = left_trial_bound + duration
+
+    from_index_array = np.searchsorted(timestamps, left_trial_bound)
+    to_index_array = np.searchsorted(timestamps, right_trial_bound)
+
+    data_list = []
+    timestamps_list = []
+    alignment_column_list = []
+    for trial_index, (idx_start, idx_stop) in enumerate(zip(from_index_array, to_index_array)):
+        trial_data = get_time_series_data(time_series, idx_start=idx_start, idx_stop=idx_stop, data_column=data_column)
+        data_list.append(trial_data)
+        timestamps_list.append(timestamps[idx_start:idx_stop])
+        alignment_column_list.append([trials_table_df.start_time[trial_index]] * (idx_stop - idx_start))
+
+    data_dict = {
+        "data": np.concatenate(data_list),
+        "timestamps": np.concatenate(timestamps_list),
+        f"{alignment_column}": np.concatenate(alignment_column_list),
+    }
+
+    data_df = pd.DataFrame(data_dict)
+
+    data_df_trialized = pd.merge(left=data_df, right=trials_table_df, on=alignment_column, how="left")
+    data_df_trialized["centered_timestamps"] = data_df_trialized.timestamps - data_df_trialized[alignment_column]
+
+    return data_df_trialized
+
+
+def create_empty_figure(text="No data"):
+    empty_figure = go.Figure()
+    empty_figure.update_layout(
+        xaxis={"visible": False},
+        yaxis={"visible": False},
+        annotations=[
+            {
+                "text": f"{text}",
+                "xref": "paper",
+                "yref": "paper",
+                "showarrow": False,
+                "font": {"size": 28},
+            }
+        ],
+    )
+
+    return empty_figure
+
+
+def calculate_moving_average_over_trials(df, moving_average_window):
+
+    df_sort = df.sort_values(by="centered_timestamps")
+    df_sort["moving_average"] = df_sort["data"].rolling(moving_average_window).mean()
+
+    return df_sort
+
+
+def add_moving_average_traces(figure, df, facet_col, facet_row):
```

Comment:
This seems specific to your widget. Maybe it would be better as a static method.